### DISCUSSION TITLE
Support for Kohler APM303 & APM403 as well as Deepsee  DSE6020KII & DSE8610

### DISCUSSION THREAD

**Discussion by AKSHAYGENMON** [jgyates/genmon#1222] [General]
I did read that in the "The following generator controllers are supported:" section that Kohlers APM604 as well as Deepsee's [7320MKII] is supported.  But I was also wondering if the solution file will also work with Kohlers APM303 and APM403 control as well as Deep Sees DSE6020KII & DSE8610 range

**Comment by jgyates** [comment]
I will answer this separately for the Kohler and Deep Sea controllers.

The the Deep See, the existing support will likely work although I am guessing that some minor changes would be needed. The support for Deep Sea controllers was created with input from a Deep Sea PDF that appears to support the 6xxx and 8xxx models, but some modification may be needed once it is tested with these controllers. I can make the changes with feedback from someone how actually is testing on one of these controllers.

For the Kohler models. It is likely I can add support for this however I do not have documentation for the APM303 and APM403. To add support, [these steps](https://github.com/jgyates/genmon/wiki/Appendix-N-Genmon-Supporting-Other-Controller-Types#the-process-of-creating-a-new-controller-support-within-genmon) would need to be followed. I can make the modfications to the JSON file but I would need modbus registers dumps and the PDF files with the modbus register definitions.

Let me know if you have any questions.

**Reply by AKSHAYGENMON** [reply]
Hi jgyates!  Thanks for your prompt and detailed reply.  Honestly I am such a newbie to all of this.  Watched HandyTVDads tutorial on Generator monitoring which made it seem so easy:  _**https://www.youtube.com/watch?v=w0DvUAiXSmU&t=1s**_.  Having said that I am not a newbie to ESP32s many-to-one communication, and interfacing arduino to RS485 digital inputs from sensors using MAX485 modules over long distance.  So I feel quite confident to procure a Pi and a Hat as well as Molex connector to give it a go.  Being able to sync my generator(s) to the user interface with gauges you designed is definitely spurring me on.

I am with you on installing the [modbusdump.py] program to my raspberry pi.  At the moment I am still trying to figure out what Pi and Hat I need from Pintsizes help feature https://cloud.pintsize.me/whattobuy.html.  I believe I have answered all the two dropdown questions that pertain to a Kohler APM303 and 403 series as they are not Generac units.  Based on my selection "No" on https://cloud.pintsize.me/whattobuy.html I get the response below: _**We may not have a product that meets your need, please contact us through facebook or our contact form to discuss your situation.**_
![what to buy](https://github.com/user-attachments/assets/f6c8437b-2495-4aa2-9e9f-ef7cf54c0102)

I have attached the PDF files for a Kolher APM303 and 403 controller but as described above just stuck on what hardware to procure to sync up to the generator.  any help to specific links for the items i would need to order would be greatly appreciated.  Once procured I will install the [modbusdump.py] to push for modbus registers dumps.  Quite confidant I should be able to figure that part out.  

Also your point about the Deep See is well noted.  I hope to attempt that next after getting the hardware ordered.  

[AMP303 User Manual.pdf](https://github.com/user-attachments/files/18232585/AMP303.User.Manual.pdf)

[APM403 User Manual_EN.pdf](https://github.com/user-attachments/files/18232582/APM403.User.Manual_EN.pdf)

**Reply by jgyates** [reply]
Unfortunatly these PDFs are user manuals, not documents that describe the modbus registers that are unique to the AMP303 an AMP403 controllers. I would need the modbus register definitions to add support for these controllers. Typically this is something that a Kohler sales rep can give you or you can download if you have access to more technical specs than a user manual.

Regarding what to buy, you have several options. If you are using RS-422 or RS-485 then I would sugest looking at a few of the project wiki entries:

https://github.com/jgyates/genmon/wiki/Appendix-R---MEBAY-Controller-with-RS-485

https://github.com/jgyates/genmon/wiki/Appendix-P-Briggs-and-Stratton-Controller-Information

If your controller uses RS-232 then the pintze.me HAT will work nicely. A lot of people who use this project use it.

Regarding setting up, this wiki page: 

https://github.com/jgyates/genmon/wiki/3.3--Setup-genmon-software

describes how to setup the software. It should be followed regardless of the serial interface. Once you install the project software as described in the link you can run both genmon and modbusdump.

**Reply by skipfire** [reply]
@AKSHAYGENMON I'm the guy behind PintSize, none of our products will work with RS485, but there are USB and HAT options on Amazon and at a lot of IoT/maker retailers for RS485 that will work fine.  I'd recommend something in the Pi 3 family as those have full size USB port (the 3A+ has one and the 3B & 3B+ both have 4).  If you have Ethernet at the generator definately go with the 3B or 3B+.  If you are running a cable extension into a climate controlled environment the Pi 4 and Pi 5 are both good options, though potentially overkill - I don't recommend them at the generator because of heat management issues.

Depending on what connections the generators have for the data output your setup may be quite simple.  If it has a DB9 port then a USB to serial cable will likely be all you would need or a RS485 HAT with a cable.  If it has screw terminals then you can buy a DB9 screwdown adapter (the genmon make a cable section has a reference to one) and that can plug into the RS485 HAT or serial cable.  Generac is a bit more of a special case because they are using Molex which is typically power cabling for a PC or used in cars so there aren't data accessories for it.

I'm happy to answer whatever questions I can, but I've never had access to a non-Generac generator so I don't know a whole lot there.

**Reply by AKSHAYGENMON** [reply]
Hi jgyates!  Thanks again for your response.  All points are well noted.  I appreciate your feedback and I see feedback from @skipfire which also gives me some additional info to get going on the hardware side.  In the meantime I was able to find a modbus pdf for the APM303 specifically.  Would this attachment be what you are looking for?  If not I am going to be meeting with a tech representative who is an authorized distrubutor of Kohler gensets in this region of subsaharan africa as that is where I am located.   I am confident I should be able to get what youre looking for from him.  I greatly appreciate your support on this.  

[417060119-APM303.pdf](https://github.com/user-attachments/files/18246455/417060119-APM303.pdf)

**Reply by AKSHAYGENMON** [reply]
Thanks for your response @skipfire.  Your feedback is well noted.  Going to review the connector types on the APM303 APM403 and DEEP SEE units.  I appreciate your support too. Will likely be picking your brains one last time before i finalize my cart and move forward with my hardware order.

**Reply by sergio1986x** [reply]
> Lamentablemente, estos PDF son manuales de usuario, no documentos que describan los registros Modbus específicos de los controladores AMP303 y AMP403. Necesitaría las definiciones de los registros Modbus para añadir compatibilidad con estos controladores. Normalmente, un representante de ventas de Kohler puede proporcionártela o puedes descargarla si tienes acceso a especificaciones técnicas más completas que las que ofrece un manual de usuario.
> 
> En cuanto a qué comprar, tienes varias opciones. Si utilizas RS-422 o RS-485, te sugiero que consultes algunas de las entradas de la wiki del proyecto:
> 
> https://github.com/jgyates/genmon/wiki/Appendix-R---MEBAY-Controller-with-RS-485
> 
> https://github.com/jgyates/genmon/wiki/Appendix-P-Briggs-and-Stratton-Controller-Information
> 
> Si tu controlador usa RS-232, el HAT de pintze.me funcionará perfectamente. Mucha gente que usa este proyecto lo usa.
> 
> Respecto a la configuración, esta página wiki:
> 
> https://github.com/jgyates/genmon/wiki/3.3--Setup-genmon-software
> 
> Describe cómo configurar el software. Debe seguirse independientemente de la interfaz serial. Una vez instalado el software del proyecto, como se describe en el enlace, podrá ejecutar tanto genmon como modbusdump.

SDMO APM303 – Modbus RTU / RS485 – Mapa completo de registros

Información recopilada del manual SDMO APM303-MODBUS, documento 33530004401, versión 2016.

Comunicación

- Protocolo: Modbus RTU
- Interfaz: RS485
- Dirección esclavo por defecto: "5"
- Velocidad: "9600 baud"
- Bits de datos: "8"
- Paridad: ninguna
- Stop bits: "1"
- APM303 funciona como esclavo
- Longitud máxima de bus indicada por SDMO: 1200 m
- Terminación recomendada: 120 Ω en ambos extremos
- Cable: STP / par trenzado apantallado

Funciones Modbus implementadas

Función| Hex| Descripción
01| "0x01"| Read Coils
02| "0x02"| Read Discrete Inputs
04| "0x04"| Read Input Registers
05| "0x05"| Write Single Coil

El APM303 NO implementa las funciones 03/06/15/16 según este manual.

---

1. ENTRADAS DIGITALES HARDWARE

Función Modbus: "02"

Dirección| Descripción| Terminal
"0x0000"| Entrada hardware 1 / Emergency Stop| T04
"0x0001"| Entrada hardware 2| T10
"0x0002"| Entrada hardware 3| T11
"0x0003"| Entrada hardware 4| T12
"0x0004"| Entrada hardware 5| T13
"0x0005"| Entrada hardware 6| T14
"0x0006"| Entrada hardware 7| T15

Las entradas T10–T15 dependen de la asignación/configuración del APM303.

---

2. SALIDAS DIGITALES HARDWARE

Se leen mediante función "02".

Dirección| Descripción| Terminal
"0x0010"| Salida hardware 1 / Starter| T05
"0x0011"| Salida hardware 2 / Fuel Solenoid| T06
"0x0012"| Salida hardware 3| T07
"0x0013"| Salida hardware 4| T08
"0x0014"| Salida hardware 5| T09

Asignación estándar:

- T07 = "O08" / Air preheating
- T08 = "O07" / Ready To Load
- T09 = "O15" / Common Shutdown

T07/T08/T09 pueden tener otra función dependiendo de la configuración.

---

3. ENTRADAS DIGITALES LÓGICAS

Función Modbus: "02"

Dirección| Función
"0x0060"| Emergency Stop
"0x0061"| Remote Start/Stop – I02
"0x0062"| Access Lock / APM303 locked – I04
"0x0063"| GCB feedback – I07
"0x0064"| External Warning 1 – I10
"0x0065"| External Warning 2 – I11
"0x0066"| External Warning 3 – I12
"0x0067"| External Shutdown 1 – I13
"0x0068"| External Shutdown 2 – I14
"0x0069"| External Shutdown 3 – I15
"0x006A"| Low Fuel Level – I20
"0x006B"| Low Oil Pressure – I22
"0x006C"| High Coolant Temperature – I24

---

4. ESTADOS / ALARMAS / DEFECTOS / SALIDAS LÓGICAS

Función Modbus: "02"

Dirección| Estado / alarma / función
"0x0020"| Starter
"0x0021"| Fuel Solenoid
"0x0022"| Stop Solenoid – O03
"0x0023"| General Alarm – O04
"0x0024"| GCB Close/Open – O05
"0x0025"| Ready To Load – O07
"0x0026"| Preheat – O08
"0x0027"| Running / grupo estabilizado
"0x0028"| Automatic Mode ON/OFF
"0x0029"| Island Operation
"0x002A"| Common Warning – O14
"0x002B"| Maintenance Required
"0x002C"| Low Battery
"0x002D"| Low Fuel Level
"0x002E"| External Warning 1
"0x002F"| External Warning 2
"0x0030"| External Warning 3
"0x0031"| Generator CCW Rotation
"0x0032"| Battery Flat
"0x0033"| Common Shutdown – O15
"0x0034"| Emergency Stop Active
"0x0035"| Overspeed
"0x0036"| Underspeed
"0x0037"| Low Oil Pressure
"0x0038"| High Coolant Temperature
"0x0039"| External Shutdown 1
"0x003A"| External Shutdown 2
"0x003B"| External Shutdown 3
"0x003C"| GCB Fail
"0x003D"| Generator Voltage High
"0x003E"| Generator Voltage Low
"0x003F"| Generator Frequency High
"0x0040"| Generator Frequency Low
"0x0041"| Start Fail
"0x0042"| Stop Fail
"0x0043"| Generator Overcurrent / Short Circuit
"0x0044"| Generator kW Overload
"0x0045"| Choke – O10
"0x0046"| Glow Plugs – O11
"0x0047"| Valve Extinguisher – O13

---

5. CONTROL REMOTO

⚠️ Estos registros son de ESCRITURA.

Función Modbus: "05 – Write Single Coil"

Dirección| Función
"0x0000"| Remote START
"0x0001"| Remote STOP
"0x0002"| Remote AUTO MODE

El APM303 utiliza estos coils para cambiar el modo de funcionamiento remotamente.

---

6. REGISTROS ANALÓGICOS / MEDICIONES

Función Modbus: "04 – Read Input Registers"

Dirección| Medición| Tipo| Escala / unidad
"0x0000"| Generator Voltage L1-N| U16| V
"0x0001"| Generator Voltage L2-N| U16| V
"0x0002"| Generator Voltage L3-N| U16| V
"0x0003"| Generator Voltage L1-L2| U16| V
"0x0004"| Generator Voltage L2-L3| U16| V
"0x0005"| Generator Voltage L1-L3| U16| V
"0x0006"| Generator Current L1| U16| A / 0.1 A
"0x0007"| Generator Current L2| U16| A / 0.1 A
"0x0008"| Generator Current L3| U16| A / 0.1 A
"0x0009"| Generator Total kW| S16| kW / 0.1 kW
"0x000A"| Generator Total kVA| S16| kVA / 0.1 kVA
"0x000B"| Generator Total PF| S16| /100
"0x000C"| Generator kW L1| S16| kW / 0.1 kW
"0x000D"| Generator kW L2| S16| kW / 0.1 kW
"0x000E"| Generator kW L3| S16| kW / 0.1 kW
"0x000F"| Generator kVA L1| S16| kVA / 0.1 kVA
"0x0010"| Generator kVA L2| S16| kVA / 0.1 kVA
"0x0011"| Generator kVA L3| S16| kVA / 0.1 kVA
"0x0012"| Generator PF L1| S8| /100
"0x0013"| Generator PF L2| S8| /100
"0x0014"| Generator PF L3| S8| /100
"0x0015"| Load Character| U8| R/L/C ASCII
"0x0016"| Engine RPM| U16| RPM
"0x0017"| Generator Frequency| U16| Hz / 0.1 Hz
"0x0018"| Power Reading Precision| U16| 0 = entero / 1 = decimal
"0x0019"| Battery Voltage| S16| V / 0.1 V
"0x001A"| BIN| U16| Estado entradas físicas
"0x001B"| BOUT| U16| Estado salidas físicas
"0x001C"| Oil Pressure| S16| depende configuración
"0x001D"| Coolant Temperature| S16| °C o °F
"0x001E"| Fuel Level| S16| %
"0x001F"| Unit System| U8| 0 = Metric / 1 = US
"0x0020"| D+ Charging Alternator| S16| V / 0.1 V
"0x0021"| Active Energy kWh – High Word| U32| kWh
"0x0022"| Active Energy kWh – Low Word| U32| kWh
"0x0023"| Maintenance Timer| U16| horas
"0x0024"| Start Counter| U16| cantidad
"0x0025"| Genset Serial Number – High Word| U32| —
"0x0026"| Genset Serial Number – Low Word| U32| —
"0x0027"| Software/Firmware Version| U16| M×100 + N
"0x0028"| Software Patch Version| U16| P×1000+B
"0x0029"| Running Hours – High Word| U32| horas
"0x002A"| Running Hours – Low Word| U32| horas
"0x002B"| Nominal Current| U16| A
"0x002C"| Connection Type| U16| 1=1Ph+N, 2=2Ph+N, 3=3Ph, 4=3Ph+N

Los registros "0x0029"–"0x002C" aparecen en la versión 2016 del documento y son especialmente importantes porque no aparecen en la documentación APM303 más antigua que circula por Internet.

---

7. ESCALAMIENTO

El parámetro "Power Reading Precision" determina si determinados valores se presentan con 0 o 1 decimal.

Ejemplos:

- Corriente "107" → "10.7 A" si escala 0.1
- Frecuencia "500" → "50.0 Hz"
- Batería "239" → "23.9 V"
- PF "95" → "0.95"

La presión de aceite depende de si el APM303 está configurado en bar o psi.

La temperatura depende de si está configurado en °C o °F.

---

8. EJEMPLO: LEER TENSIÓN DE BATERÍA

Para leer "0x0019":

Slave:       05
Function:    04
Register:    0019
Quantity:    0001
CRC:         XXXX

Trama:

05 04 00 19 00 01 XX XX

Si el APM303 devuelve:

05 04 02 00 EF XX XX

"0x00EF" = 239 decimal.

Escala = 0.1 V.

Resultado:

23.9 V

---

9. EJEMPLO: LEER READY TO LOAD

Dirección:

0x0025

Función:

02

Cantidad:

1 bit

Petición:

05 02 00 25 00 01 XX XX

---

10. EJEMPLO: LEER COMMON SHUTDOWN

Dirección:

0x0033

Función:

02

Petición:

05 02 00 33 00 01 XX XX

---

11. EJEMPLO: COMANDO START

⚠️ PRECAUCIÓN: esto es una escritura y puede provocar el arranque del grupo.

Función:

05

Coil:

0x0000

START ON:

FF 00

START OFF:

00 00

La respuesta normalmente es un eco de la petición.

---

12. IMPORTANTE SOBRE LAS DIRECCIONES

Las direcciones indicadas aquí son las direcciones Modbus del documento SDMO en hexadecimal.

Por ejemplo:

0x0019 = 25 decimal

No hay que confundirlas automáticamente con direcciones tipo:

30026
40026

dependiendo del software Modbus utilizado.

El manual utiliza directamente la dirección Modbus hexadecimal.

---

13. DOCUMENTACIÓN

Manual SDMO:

APM303-MODBUS – 33530004401

Versión publicada en 2016.

El documento indica que es propiedad de SDMO Industries y describe específicamente la comunicación RS485/Modbus del APM303.

También existe una versión anterior:

APM303-MODBUS – 33530004501_1_1 – 21/01/2015

que contiene el mapa original de registros.

La discusión del foro donde otro usuario adjuntó el PDF específico de APM303 Modbus está aquí:

https://forum.logicmachine.net/showthread.php?tid=4845

**Comment by sergio1986x** [comment]
Hi, I saw your discussion about the Kohler APM303/APM403. I'm also working with APM303 controllers and I'm trying to obtain the original SDMO APM PC Suite / APM303 Configurator software and access to GAIA. You mentioned that you were in contact with an authorized Kohler distributor technician. Were you able to obtain the software or contact information for that technician? Any help would be greatly appreciated.

**Reply by jgyates** [reply]
Hi @sergio1986x 

I do not have any direct or indirect contact with Kohler. I have no access to any configurator software. Someone else in the forum may be able to put you in contact with Kohler but I do not have that info.
